### PR TITLE
Feat(releases): Ability to schedule a release 

### PR DESCRIPTION
**Pull request by madhurisandbhor**
### What does it do?

- Adds the ability to schedule a release
- Timezone selected will default to the users location

### How to test it?
When creating or editing a release, you should see a modal to edit the name and the “Schedule release” checkbox should be available:
- When schedule release is checked, you see 3 fields below: date, time and timezone. User should not be able to save unless all fields are selected.
- When schedule release is unchecked, you don’t see scheduling fields and user should be able to save.

### CODE REVIEW

File: packages/core/content-releases/admin/src/components/ReleaseModal.tsx [hunk 1]
```diff
@@ -51,24 +95,130 @@ export const ReleaseModal = ({
...
+                            id: 'content-releases.modal.form.input.label.date',
+                            defaultMessage: 'Date',
+                          })}
+                          name="date"
+                          onChange={(date) => {
+                            const isoFormatDate = date
+                              ? formatISO(date, { representation: 'date' })
+                              : null;
+                            setFieldValue('date', isoFormatDate);
+                          }}
```
Comment (joshuaellis) [inline packages/core/content-releases/admin/src/components/ReleaseModal.tsx:165]: Can `Formik` not handle Dates as values? I'm wondering if you convert it to what the API needs before posting?

File: packages/core/content-releases/admin/src/components/ReleaseModal.tsx [hunk 2]
```diff
@@ -80,7 +230,11 @@ export const ReleaseModal = ({
                 <Button
                   name="submit"
                   loading={isLoading}
-                  disabled={!values.name || values.name === initialValues.name}
+                  disabled={
+                    !dirty ||
+                    !values.name ||
+                    (values.isScheduled && (!values.time || !values.date || !values.timezone))
```
Comment (joshuaellis) [inline packages/core/content-releases/admin/src/components/ReleaseModal.tsx:236]: i'm not sure about this to be honest, i think it's quite bad UX because user's can submit their form and find out what the issue is, they're just gonna be confused. Why are we not using the validation to give feedback?

File: packages/core/content-releases/admin/src/components/tests/ReleaseModal.test.tsx [hunk 3]
```diff
@@ -67,4 +68,94 @@ describe('ReleaseModal', () => {
...
+    const handleCloseMocked = jest.fn();
+    render(
+      <ReleaseModal
+        handleClose={handleCloseMocked}
+        handleSubmit={jest.fn()}
+        initialValues={{ name: 'title', date: null, time: '', timezone: '', scheduledAt: null }}
+        isLoading={false}
+      />
+    );
+    const dialogContainer = screen.getByRole('dialog');
```
Comment (joshuaellis) [inline packages/core/content-releases/admin/src/components/tests/ReleaseModal.test.tsx:82]: Isn't the only thing you render a dialog? Why do you need to keep doing `within`? 🤔

File: packages/core/content-releases/admin/src/components/tests/ReleaseModal.test.tsx [hunk 4]
```diff
@@ -67,4 +68,94 @@ describe('ReleaseModal', () => {
...
+    const scheduleReleaseCheck = within(dialogContainer).getByRole('checkbox', {
+      name: /schedule release/i,
+    });
+
+    // Schedule release checkbox is not checked and date field is not visible
+    expect(scheduleReleaseCheck).not.toBeChecked();
+    const date = within(dialogContainer).queryByRole('combobox', {
+      name: /date/i,
```
Comment (joshuaellis) [inline packages/core/content-releases/admin/src/components/tests/ReleaseModal.test.tsx:90]: is there a reason you're using Regex to find the labels? You could use the actual names?

File: packages/core/content-releases/admin/src/pages/ReleaseDetailsPage.tsx [hunk 5]
```diff
@@ -832,12 +832,25 @@ const ReleaseDetailsPage = () => {
...
+  const title = releaseData?.name || '';
+  const scheduledAt = releaseData?.scheduledAt ? new Date(releaseData.scheduledAt) : null;
+  const timezone = releaseData?.timezone;
+
+  let time = '';
+  if (scheduledAt) {
+    const hours = scheduledAt.getHours();
+    const minutes = scheduledAt.getMinutes();
+    time = `${hours.toString().padStart(2, '0')}:${minutes.toString().padStart(2, '0')}`;
+  }
```
Comment (joshuaellis) [inline packages/core/content-releases/admin/src/pages/ReleaseDetailsPage.tsx:846]: I think you should be using something like:

https://github.com/strapi/design-system/blob/c4ecb525c730cd100c6a7a1fe03d7dabea759e9b/packages/strapi-design-system/src/hooks/useDateFormatter.ts#L32

to handle this sort of stuff, string manipulation is very messy and prone to errors.

File: packages/core/content-releases/admin/src/components/ReleaseModal.tsx [hunk 6]
```diff
@@ -34,6 +49,35 @@ export const ReleaseModal = ({
...
+    const combinedTime = combinedDateTime.getTime();
+    const combinedTimeWithOffset = combinedTime - offset * 60 * 60 * 1000; // Convert offset in hours to milliseconds
+
+    // Because new Date always adds local timezone offset, remove it to set the correct UTC time
+    const dateObject = new Date(combinedTimeWithOffset);
+    const scheduledDate = new Date(
+      dateObject.getTime() - dateObject.getTimezoneOffset() * 60 * 1000
+    );
+    return scheduledDate;
+  };
```
Comment (joshuaellis) [inline packages/core/content-releases/admin/src/components/ReleaseModal.tsx:80]: I think you should be doing something similar to how we do the TimePicker in the DS for string manipulation, this is prone to bugs and the separator isn't guaranteed to be `:` because it comes from the web APIs stuff:

https://github.com/strapi/design-system/blob/8f6f2cf7d047e6cad4b17270f1950fd9426e29d7/packages/strapi-design-system/src/TimePicker/TimePicker.tsx#L88

File: packages/core/content-releases/admin/src/components/ReleaseModal.tsx [hunk 7]
```diff
@@ -51,24 +95,130 @@ export const ReleaseModal = ({
...
+                  onChange={(event) => {
+                    setFieldValue('isScheduled', event.target.checked);
+                    if (!event.target.checked) {
+                      // Clear scheduling info from a release on unchecking schedule release
+                      setFieldValue('date', null);
+                      setFieldValue('time', '');
+                    } else {
+                      setFieldValue('date', initialValues.date);
+                      setFieldValue('time', initialValues.time);
+                    }
```
Comment (joshuaellis) [inline packages/core/content-releases/admin/src/components/ReleaseModal.tsx:137]: I find this a bit weird, why are doing this here?

**Review by joshuaellis — CHANGES_REQUESTED** [review]
I think this needs some tidying up before we merge it. Also should it's ability to show in the form also be dictated if the future flag is enabled similar to it's server counterpart? Otherwise we're gonna run into a lot of confusion 🤔

File: packages/core/content-releases/admin/src/components/ReleaseModal.tsx (hunk 7 above)
Comment (madhurisandbhor) [reply packages/core/content-releases/admin/src/components/ReleaseModal.tsx]: On ticking/unticking I need to keep track of dirty state and so refill to initial data if ticked back to its original state.
https://github.com/strapi/strapi/assets/7886244/afa7c5ed-b288-4d7f-9b9d-aec523be486f

Comment (madhurisandbhor) [reply packages/core/content-releases/admin/src/components/ReleaseModal.tsx]: https://github.com/strapi/strapi/assets/7886244/72d9f7b6-26c3-4b98-a7b0-2c0d7f7c8883

File: packages/core/content-releases/admin/src/components/tests/ReleaseModal.test.tsx (hunk 4 above)
Comment (madhurisandbhor) [reply packages/core/content-releases/admin/src/components/tests/ReleaseModal.test.tsx:99]: no, just a habit.

File: packages/core/content-releases/admin/src/components/ReleaseModal.tsx (hunk 2 above)
Comment (madhurisandbhor) [reply packages/core/content-releases/admin/src/components/ReleaseModal.tsx:225]: I have same opinion as you, but last time I implemented error state and I was asked to revert it.
https://www.figma.com/file/MWPKHNT98LLWHGRp3MyWqu?node-id=1339:17473&mode=design#643292761

Comment (madhurisandbhor) [reply packages/core/content-releases/admin/src/components/ReleaseModal.tsx:225]: On the contrary, when I try to use error state on these fields which are all select DD, its of no use as inputs are always going to be blank on wrong entry, and so we disable Save button.

File: packages/core/content-releases/admin/src/components/ReleaseModal.tsx (hunk 1 above)
Comment (madhurisandbhor) [reply packages/core/content-releases/admin/src/components/ReleaseModal.tsx]: No, it didnt work.

File: packages/core/content-releases/admin/src/components/ReleaseModal.tsx (hunk 2 above)
Comment (Feranchz) [reply packages/core/content-releases/admin/src/components/ReleaseModal.tsx:225]: Based on what discussed on Slack I think you can safely remove this 🎉

File: packages/core/content-releases/admin/src/components/ReleaseModal.tsx [hunk 8]
```diff
@@ -99,3 +242,50 @@ export const ReleaseModal = ({
...
+     * So we need to append the 0's.
+     */
+    if (!utcOffset.includes('+') && !utcOffset.includes('-')) {
+      utcOffset = `${utcOffset}+00:00`;
+    }
+
+    return { offset: utcOffset, value: timezone } satisfies ITimezoneOption;
+  });
+
+  const currentTimezone = timezoneList.find(
```
Comment (Feranchz) [inline packages/core/content-releases/admin/src/components/ReleaseModal.tsx:286]: Maybe I'm wrong but I was struggling for a while thinking that this value refers to the selected timezone, maybe we could change this to `systemTimezone` to be more precise?

File: packages/core/content-releases/admin/src/pages/ReleasesPage.tsx [hunk 9]
```diff
@@ -141,6 +141,12 @@ const StyledAlert = styled(Alert)`
 
 const INITIAL_FORM_VALUES = {
   name: '',
+  date: null,
+  time: '',
+  // Remove future flag check after Scheduling Beta release and replace with true as creating new release should include scheduling by default
+  isScheduled: window.strapi.future.isEnabled('contentReleasesScheduling'),
```
Comment (Feranchz) [inline packages/core/content-releases/admin/src/pages/ReleasesPage.tsx:147]: I think this is not needed here, `isScheduled` when creating a release should be false always

File: packages/core/content-releases/admin/src/components/ReleaseModal.tsx [hunk 10]
```diff
@@ -99,3 +242,50 @@ export const ReleaseModal = ({
...
+  const timezoneList: ITimezoneOption[] = Intl.supportedValuesOf('timeZone').map((timezone) => {
+    /**
+     * This will be in the format GMT${OFFSET} where offset could be
+     * nothing, a four digit string e.g. +05:00 or -08:00
+     */
+    const offsetPart = new Intl.DateTimeFormat('en', {
+      timeZone: timezone,
+      timeZoneName: 'longOffset',
+    })
+      .formatToParts()
```
Comment (Feranchz) [inline packages/core/content-releases/admin/src/components/ReleaseModal.tsx:265]: Here is an interesting case, if you use `.formatToParts()` like that it would use the today's date to get the timezone offset. That means that for me is going to show "UTC+01:00 Europe/Madrid" 

But what if I selecting a date for July? Above you are using `.zonedTimeToUtc` to parse the date to UTC0 and it handles this case, I mean, considering I'm using Europe/Madrid as timezone and I select a date and hour in July, it would parse that date into UTC subtracting -2 to the time. But on my Timezone input I would see "UTC+01:00 Europe/Madrid", maybe this could cause confusion?

If you pass the selected date to `.formatToParts()` it would give you the offset based on that, so if I select a date on July in my dropdown I would see "UTC+02:00 Europe/Madrid"

wdyt? 

cc @joshuaellis

**Review by Feranchz — COMMENTED** [review]
It's looking really good @madhurisandbhor, thank you for this! I did QA and can't find any problem, I just let some comments to have your opinion

File: packages/core/content-releases/admin/src/components/ReleaseModal.tsx [hunk 11]
```diff
@@ -34,6 +50,18 @@ export const ReleaseModal = ({
...
+  // Set default first value from the timezone list with offset UTC+00:00
+  const { timezoneList, currentTimezone = { value: 'Africa/Abidjan' } } = useTimezone();
+  const IsSchedulingEnabled = window.strapi.future.isEnabled('contentReleasesScheduling');
+  /**
+   * Generate scheduled time using selected date, time and timezone
+   */
+  const getScheduledTimestamp = (values: FormValues) => {
+    const { date, time, timezone } = values;
+    if (!date || !time || !timezone) return null;
+    const formattedDate = parse(`${time}`, 'HH:mm', new Date(date));
```
[the remark this answers is not in the document]
Comment (madhurisandbhor) [reply packages/core/content-releases/admin/src/components/ReleaseModal.tsx:62]: Right! I had date as well so..

File: packages/core/content-releases/admin/src/pages/ReleasesPage.tsx (hunk 9 above)
Comment (madhurisandbhor) [reply packages/core/content-releases/admin/src/pages/ReleasesPage.tsx:147]: No, I had confirmed this with Yannis and so updated it

File: packages/core/content-releases/admin/src/components/ReleaseModal.tsx (hunk 10 above)
Comment (madhurisandbhor) [reply packages/core/content-releases/admin/src/components/ReleaseModal.tsx:265]: Indeed, thats very important param I missed.

File: packages/core/content-releases/admin/src/components/ReleaseModal.tsx [hunk 12]
```diff
@@ -99,3 +237,90 @@ export const ReleaseModal = ({
...
+    // We want to show time based on UTC, not GMT so we swap that.
+    let utcOffset = offset.replace('GMT', 'UTC');
+
+    // For perfect UTC (UTC+0:00) we only get the string UTC, So we need to append the 0's.
+    if (!utcOffset.includes('+') && !utcOffset.includes('-')) {
+      utcOffset = `${utcOffset}+00:00`;
+    }
+
+    // Offset and timezone are concatenated with '-', so to split and save the required timezone in DB
+    return { offset: utcOffset, value: `${utcOffset}-${timezone}` } satisfies ITimezoneOption;
```
Comment (madhurisandbhor) [inline packages/core/content-releases/admin/src/components/ReleaseModal.tsx:271]: I have updated the code so the changes made in date reflects in timezone. 
Though I faced 1 issue to update the combobox value, so I changed value which is `<timezone>`(Europe/Paris) to `<offset-timezone>`(UTC+01:00-Europe/Paris) and `textValue` prop added on combobox.

**Review by Feranchz — APPROVED** [review]
👍 🚀

File: packages/core/content-releases/server/src/controllers/validation/release.ts [hunk 13]
```diff
@@ -1,4 +1,4 @@
 import { validateYupSchema } from '@strapi/utils';
 import { RELEASE_SCHEMA } from '../../../../shared/validation-schemas';
 
-export const validateRelease = validateYupSchema(RELEASE_SCHEMA);
+export const validateRelease = validateYupSchema(RELEASE_SCHEMA.noUnknown());
```
Comment (joshuaellis) [inline packages/core/content-releases/server/src/controllers/validation/release.ts:4]: why is this added here and not on the actual schema where you also removed it prior? :)

Comment (madhurisandbhor) [reply packages/core/content-releases/server/src/controllers/validation/release.ts:4]: It was not working as all fields were not passed to BE, but its not the case now so reverted.

File: packages/core/content-releases/admin/src/components/tests/ReleaseModal.test.tsx [hunk 14]
```diff
@@ -1,70 +1,89 @@
...
+    await waitFor(() => {
+      expect(time).toBeInTheDocument();
+    });
+
+    const timezone = screen.getByRole('combobox', {
+      name: /timezone/i,
+    });
+    await waitFor(() => {
+      expect(timezone).toBeInTheDocument();
+    });
```
[the remark this answers is not in the document]
Comment (madhurisandbhor) [reply packages/core/content-releases/admin/src/components/tests/ReleaseModal.test.tsx:87]: right!

**Review by joshuaellis — APPROVED** [review]
Great work. I wonder if we could have avoided date-fns with the web APIs but either way, I think it's a great clear solution 💯
